hs_core/views/resource_rest_api.py

```diff
@@ -0,0 +1,622 @@
+from rest_framework.views import APIView
+from rest_framework.response import Response
+from rest_framework import status, generics
+from rest_framework.exceptions import *
+from hs_core import hydroshare
+from hs_core.hydroshare import file_size_limit_for_display
+from hs_core.views import utils
+import simplejson as sj
+import serializers
+from hs_core.signals import *
+import os
+
+class ResourceRetrieveUpdateDelete(APIView):
+    """
+    Retrieve, update or delete a resource
+
+    REST URL: hsapi/resource/{pk}
+
+    HTTP method: GET
+
+    HTTP status code (on success):200
+
+    :return: (on success): resource bag file if request content-type is 'application/zip' otherwise resource science metadata file
+
+    REST URL: hsapi/resource/{pk}
+
+    HTTP method: DELETE
+
+    HTTP status code (on success):200
+
+    :return: (on success): json string of the format: {'resource_id':pk}
+
+    REST URL: hsapi/resource/{pk}
+
+    HTTP method: PUT
+
+    HTTP status code (on success):200
+
+    :return: (on success): json string of the format: {'resource_id':pk}
+
+    :type   str
+
+    :param  pk: resource id
+
+    :rtype:  json string for http methods DELETE and PUT, and resource file data bytes for GET
+
+    :raises:
+
+
+    &nbsp;&nbsp;&nbsp;NotFound: return json format: {'detail': 'No resource was found for resource id':pk}
+
+    &nbsp;&nbsp;&nbsp;HTTP status code: 404
+
+    &nbsp;&nbsp;&nbsp;PermissionDenied: return json format: {'detail': 'You do not have permission to perform this action.'}
+
+    &nbsp;&nbsp;&nbsp;HTTP status code:403
+
+    &nbsp;&nbsp;&nbsp;ValidationError: return json format: {parameter-1': ['error message-1'], 'parameter-2': ['error message-2'], .. }
+
+    &nbsp;&nbsp;&nbsp;HTTP status code:400
+    """
+
+    @property
+    def allowed_methods(self):
+        return ['GET', 'PUT', 'DELETE']
+
+    def get(self, request, pk):
+        # check the user is authorized for this action and if the resource exists for the given resource id (pk)
+        utils.authorize(request, pk, view=True)
+        resource_bag = hydroshare.get_resource(pk)
+
+        if request.content_type == 'application/zip':
+            # redirects to django_irods/views.download function
+            return HttpResponseRedirect(resource_bag.bag.url)
+        else:
+            return redirect('science_metadata_get_update', pk=pk)
+
+    def put(self, request, pk):
+        # check the user is authorized for this action and if the resource exists for the given resource id (pk)
+        utils.authorize(request, pk, edit=True)
+
+        resource_update_request_validator = serializers.ResourceUpdateRequestValidator(data=request.data)
+        if not resource_update_request_validator.is_valid():
+            raise ValidationError(detail=resource_update_request_validator.errors)
+
+        # TODO: update resource - not sure what exactly we need to update
+        raise NotImplementedError()
```

Comment:
Yes. I am actually leery of allowing apps to recreate the content of a Bag.
But here is a use case that would require it - another time series tool
that is used to edit outliers in a data set before the dataset is
published. The user could launch the time series editor, remove outliers,
and then re-PUT the contents of the bag - now the edited time series. In
this case, since she hasn't published it yet, she may simply want to use
the editing capabilities to edit the contents of the bag before publishing
it.

On Wed, Apr 22, 2015 at 1:06 PM Brian Miles <notifications@github.com>
wrote:

> In hs_core/views/resource_rest_api.py
> <https://github.com/hydroshare/hydroshare/pull/313#discussion_r28904387>:
>
> > +        if request.content_type == 'application/zip':
> > +            # redirects to django_irods/views.download function
> > +            return HttpResponseRedirect(resource_bag.bag.url)
> > +        else:
> > +            return redirect('science_metadata_get_update', pk=pk)
> > +
> > +    def put(self, request, pk):
> > +        # check the user is authorized for this action and if the resource exists for the given resource id (pk)
> > +        utils.authorize(request, pk, edit=True)
> > +
> > +        resource_update_request_validator = serializers.ResourceUpdateRequestValidator(data=request.data)
> > +        if not resource_update_request_validator.is_valid():
> > +            raise ValidationError(detail=resource_update_request_validator.errors)
> > +
> > +        # TODO: update resource - not sure what exactly we need to update
> > +        raise NotImplementedError()
>
> @danames <https://github.com/danames> Thanks for the specific example,
> this helps. The current API does allow for resource creation (i.e. POST
> /resource). PUT /resource/{PID} would only be used to modify an existing
> resource by specify a new BagIt archive to overwrite all resource data and
> metadata. It sounds like we don't need to allow for such wholesale changes
> to support the use case you are describing. Does this seem accurate to you?
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/hydroshare/hydroshare/pull/313/files#r28904387>.
>